### PR TITLE
[Bugfix] Add redis-py >= 4.2.0 support

### PR DESCRIPTION
**Pull request by laggardkernel**
2022.05.12: switched backend from aioredis` to `redis-py >= 4.2.0`.

---

**Update**: 

- To test with `aioredis` 2.0.0 automatically. Travis and tox configurations need to be updated. 
- Fixes aio-libs/aiocache#543

## What do these changes do?

aioredis 2.0.0 released with breaking API change. Update current backend implementation of `airedis/backends/redis.py`.

Breaking aioredis changes encounter during making this PR. aio-libs/aioredis-py#1109

## Are there changes in behavior for the user?

Nope. Lower level redis backend change. No public API affected.

## Checklist

- [x] I think the code is well written
- [x] Unit tests for the changes exist
- [ ] Documentation reflects the changes
- [ ] Add a new news fragment into the `CHANGES` folder
  * name it `<issue_id>.<type>` (e.g. `588.bugfix`)
  * if you don't have an `issue_id` change it to the pr id after creating the PR
  * ensure type is one of the following:
    * `.feature`: Signifying a new feature.
    * `.bugfix`: Signifying a bug fix.
    * `.doc`: Signifying a documentation improvement.
    * `.removal`: Signifying a deprecation or removal of public API.
    * `.misc`: A ticket has been closed, but it is not of interest to users.
  * Make sure to use full sentences with correct case and punctuation, for example: `Fix issue with non-ascii contents in doctest text files.`

### CODE REVIEW

**Comment by gerardcl** [comment]
hi @laggardkernel the PR LGTM! I like it! 👍  let me know if you want me to help on anything missing from the list! thanks a ton

**Comment by markzporter** [comment]
Would love to see this merged

**Comment by JimNero009** [comment]
Is this project still maintained? How can we as a group go about getting this merged? For us, this is blocking the migration of some keen souls to use Python 3.10 in the company.

**Comment by Boring-Mind** [comment]
> Is this project still maintained? How can we as a group go about getting this merged? For us, this is blocking the migration of some keen souls to use Python 3.10 in the company.

You can create a fork from this project and make any adjustments you want. It's pretty easy. If you need to download the package from pip, please upload it to pypi.

**Comment by JimNero009** [comment]
We may indeed have to go down that route if there is no chance of this change getting into the codebase any time soon -- is this the case? The changes here seem functionally complete? 

In other words, is it 'official' that this codebase is no longer maintained?

**Comment by argaen** [comment]
Hi, sadly I'm not maintaining this project anymore but I'm happy to invite people to the project in case anyone wants to volunteer.

**Comment by pboers1988** [comment]
@argaen has anybody volunteerd as of yet?

**Comment by Dreamsorcerer** [comment]
If anyone wants to make a start, feel free to start PRs for migrating the CI to Github. See: https://github.com/aio-libs/aiocache/issues/536#issuecomment-1024448383

**Comment by Dreamsorcerer** [comment]
I think keeping up the multi-version support is unsustainable and has little benefit. I'd like to just set the minimum version to `aioredis>=2` for the next release and keep things simple. If someone has time to update this PR to only use v2, that would be great. Otherwise, I'll look at updating later.

**Comment by laggardkernel** [comment]
@Dreamsorcerer Okay, I'd like to have a try to adapt the PR with only `aioredis>=2` support when I have spare time.

**Comment by Mark90** [comment]
@laggardkernel Did you get blocked on something, or just didn't get around to this? Otherwise, me (or someone else) could try to pick this up.

**Comment by Dreamsorcerer** [comment]
Yes, might as well move to that library then. Though it's the same work, it looks like they've merged aioredis upstream, so the only difference currently is the import statement.

**Comment by laggardkernel** [comment]
WIP: Updated the PR with only aioredis 2.x support. The test is fixed and passed, but the pytest fixtures may need some updates. I'll check that later.

> Would you agree it makes more sense to refactor aiocache to use redis-py than to make it ready for aioredis-py 2.x when the latter is likely to be deprecated?

I have read both `aioredis-py` and `redis-py` before. The last version of `redis-py` I read is still 4.1. I'll check the latest release of redis-py soon.

---

@Mark90 The time when I opened the PR I was unemployed. I had a good time browsing and reading all kinds of Python libraries, learning how they are designed. Now i'm back to work and kind of busy. Sorry. 🙇‍♂️

**Comment by laggardkernel** [comment]
aioredis 2.x only support is done.

This PR is firstly made with backward compatiblity taken into consideration.
It supported for both aioredis 1.x and 2.x.
After the author agreed to drop support for aioredis 1.x, I went into a wrong
direction that kept the old design of redis cache backend, which managed
conn and pool manually. This is required when integrating with aioredis 1.x
but not with 2.x

Luckily, I realised it after a while, and refactored the code to be similar
with the memecached cache backend. It's much simpler and much more elegant now.

---

Most of my time is not spent on refactoring the redis cache backend code.
It's taken to make a correct mock of the `aioredis.Redis` instance.
FYI, methods like `.get()` `.set()` on `Redis` are not coroutine. In fact,
when we do `await redis.get()` we're await its returned value, which is a coroutine.

Autospeccing `aioredis.Redis` doesn't exactly reflect the above behavior.
I mocked the `.get()`, `.set()` methods with `AsyncMock` now to pass the tests.
Is there a better way to mock the methods more precisely?
I guess it doesn't matter, cause all we wanna do is to ensure `Redis.method` gets called.
Whether `Redis.method` communicates with the redis server correctly is tested in `aioredis`.

Sorry, I haven't had time to check async support in `redis-py` yet.

**Comment by laggardkernel** [comment]
Done. 

The `redis >= 4.2.0` async support is based on `aioredis` and the APIs are compatible. I switched to from `aioredis` to `redis` easily. All redis related tests are fixed and passed on my local machine.

There're some breaking changes compared with `aioredis 1.x`, I'll detail them as code comment conversions.

**Comment by laggardkernel** [comment]
Sorry, I thought the pre-commit hooks was enough to detect the linting err. Maybe some of my commits were made without git hooks installed yet.

Anyway, I fixed the linting err with `make lint`. I re-disabled the performance tests. Haven't dig it deeper to fix them yet. Not only redis, there're also some problems with the memcached performance tests as well, during the test running on my local machine.

BTW, I rebased the PR after the "test key refactoring" merged today.

**Comment by joeblackwaslike** [comment]
@laggardkernel @argaen Do you guys need someone to take this PR over the finish line?  If you want to add me as a maintainer I can finish any remaining tasks necessary and cut a new release.  Thx

**Comment by laggardkernel** [comment]
The PR is completed. Just need another run of the test by the CI, which is not triggered automatically. Part of the performance test is disabled, cause it doesn't fit the connection pool in `redis-py`.

@joeblackwaslike Glad about the help you offered. I'm afraid there's nothing we could do. Just the project lacks maintenance. If you need to use aioredis in your own project, you'd better fork the repo, merge this PR and build a package by yourself.

**Comment by argaen** [comment]
Hi, as mentioned above, I'm not maintaining this project anymore but I'm happy to invite any maintainers who would like to contribute. @laggardkernel @joeblackwaslike let me know and I'll invite you

**Comment by Krukov** [comment]
Hi there,

If someone here looking for alternative with more up to date support:

https://github.com/Krukov/cashews  - ready to be used and contribute
 - using redis-py
 - python 3.7-3.11

**Comment by Dreamsorcerer** [comment]
I'll refresh myself on where this is at in a couple of days.
Tests probably just need to have master merged in order to run, maybe a force-push messed it up or something.

File: aiocache/backends/redis.py [hunk 1]
```diff
@@ -52,163 +39,171 @@ def __init__(
...
-        self.pool_min_size = int(pool_min_size)
-        self.pool_max_size = int(pool_max_size)
+        # NOTE: In redis and aioredis 2.x, ConnectionPool raises ConnectionError
+        #  but doesn't wait for connection reuse when number of conns
+        #  exceeds the limit. While in aioredis 1.x, it waits asynchronously
+        #  via asyncio.Condition. So max pool size defaults to None in redis.
+        if pool_max_size is None:
+            self.pool_max_size = None
+        else:
+            self.pool_max_size = int(pool_max_size)
```
Comment (Dreamsorcerer) [inline aiocache/backends/redis.py:71]:
```suggestion
        self.pool_max_size = None if pool_max_size is None else int(pool_max_size)
```

Comment (Dreamsorcerer) [reply aiocache/backends/redis.py:71]: We should probably just look at adding type annotations and then not doing this `int()` conversion...

File: aiocache/backends/redis.py [hunk 2]
```diff
@@ -52,163 +39,171 @@ def __init__(
...
         self.endpoint = endpoint
         self.port = int(port)
         self.db = int(db)
         self.password = password
-        self.pool_min_size = int(pool_min_size)
-        self.pool_max_size = int(pool_max_size)
+        # NOTE: In redis and aioredis 2.x, ConnectionPool raises ConnectionError
+        #  but doesn't wait for connection reuse when number of conns
+        #  exceeds the limit. While in aioredis 1.x, it waits asynchronously
+        #  via asyncio.Condition. So max pool size defaults to None in redis.
```
Comment (Dreamsorcerer) [inline aiocache/backends/redis.py:67]:
```suggestion
```

Nice to know why the change in the PR, but probably doesn't need to be committed to the code.

File: aiocache/backends/redis.py [hunk 3]
```diff
@@ -52,163 +39,171 @@ def __init__(
...
+            # NOTE: decoding can't be controlled on API level after switching to
+            #  redis, we need to disable decoding on global/connection level.
+            #  Cause some of the values are save as bytes directly, like pickle
+            #  serialized values, which may raise exc when decoded with 'utf-8'.
+            "decode_responses": False,
+            # parameter names are different
+            "socket_connect_timeout": self.create_connection_timeout,
+            "max_connections": self.pool_max_size,
+        }
+        self.client = redis.Redis(**redis_kwargs)
```
Comment (Dreamsorcerer) [inline aiocache/backends/redis.py:90]: Better to put the kwargs in directly. You lose typing information with a dict, so mypy is unlikely to catch wrong types getting passed in (e.g. if upstream API changes).

File: aiocache/backends/redis.py [hunk 4]
```diff
@@ -52,163 +39,171 @@ def __init__(
...
-    @conn
     async def _cas(self, key, value, token, ttl=None, _conn=None):
         args = [value, token]
         if ttl is not None:
             if isinstance(ttl, float):
                 args += ["PX", int(ttl * 1000)]
             else:
                 args += ["EX", ttl]
-        res = await self._raw("eval", self.CAS_SCRIPT, [key], args, _conn=_conn)
+        args = [key] + args
```
Comment (Dreamsorcerer) [inline aiocache/backends/redis.py:127]: Why not just add key to the start of the list on line 121?

File: aiocache/backends/redis.py [hunk 5]
```diff
@@ -52,163 +39,171 @@ def __init__(
...
+        return await self.client.setex(key, ttl, value)
 
-    @conn
     async def _cas(self, key, value, token, ttl=None, _conn=None):
         args = [value, token]
         if ttl is not None:
             if isinstance(ttl, float):
                 args += ["PX", int(ttl * 1000)]
             else:
                 args += ["EX", ttl]
```
Comment (Dreamsorcerer) [inline aiocache/backends/redis.py:126]: On an unrelated point, these should probably be `args.extend(...)`, rather than doing concatenation.

File: tests/performance/conftest.py [hunk 6]
```diff
@@ -1,21 +1,26 @@
...
-# TODO: Update aioredis and fix tests.
+
+# TODO:
+#  1. concurrency_error_rates test doesn't work as expected with redis-py,
+#  where both ConnectionPool and BlockingConnectionPool raise ConnectionError
+#  but don't wait for connection reuse when number of conns exceeds the limit.
+#  That's why pool in redis-py uses an unlimit number connections by default.
+#  So no HTTP req should fails in the concurrency_error_rates test with redis-py.
+#  2. On my local machine, test_memcached_getsetdel() fails, it doesn't reach
+#  the performance target.
```
Comment (Dreamsorcerer) [inline tests/performance/conftest.py:13]: Let's move this into an issue if we're not fixing it immediately.

File: tests/ut/backends/test_redis.py [hunk 7]
```diff
@@ -1,319 +1,268 @@
...
-
 
 @pytest.fixture
-def redis_pool(redis_connection):
-    class FakePool:
-        def __await__(self):
-            yield
-            return redis_connection
+def redis_pipeline():
+    pipeline = AsyncMock(name="redis_pipeline")
```
Comment (Dreamsorcerer) [inline tests/ut/backends/test_redis.py:14]: Can you try with `create_autospec()`? Code would be more robust, and the next few lines can probably be removed as well.

File: tests/ut/backends/test_redis.py [hunk 8]
```diff
@@ -1,319 +1,268 @@
...
-        )
-
-    @pytest.mark.asyncio
-    async def test_get(self, redis, redis_connection):
-        await redis._get(Keys.KEY)
-        redis_connection.get.assert_called_with(Keys.KEY, encoding="utf-8")
+    async def test_get(self, redis):
+        redis.client.get.return_value = b"value"
+        assert await redis._get(Keys.KEY) == "value"
+        redis.client.get.assert_called_with(Keys.KEY)
```
Comment (Dreamsorcerer) [inline tests/ut/backends/test_redis.py:108]: The original test use `.assert_called_with(Keys.KEY, encoding="utf-8")`?

File: tests/ut/backends/test_redis.py (hunk 7 above)
Comment (Dreamsorcerer) [reply tests/ut/backends/test_redis.py:14]: This is also on my list of tasks (https://github.com/aio-libs/aiocache/issues/536#issuecomment-1068390586), so can be looked at separately if you've not got the time.

**Comment by Dreamsorcerer** [comment]
OK, looking good. Just some minor niggles I'd like resolved.
Sorry it's taken so long.

File: tests/ut/backends/test_redis.py (hunk 8 above)
Comment (Dreamsorcerer) [reply tests/ut/backends/test_redis.py:108]: Because encoding is done in aiocache now.
